### DISCUSSION TITLE
Modular widgets

### DISCUSSION THREAD

**Discussion by nxy7** [Aylur/ags#297] [Ideas]
Hello, I've spent few weeks using AGS and I think it's pretty good abstraction over GJS. There's one thing I don't really like about it though - sharing configs with others. I'm currently running slightly modified config of @Aylur and while copying someone's config is probably not the best way to use something like AGS, building your own config from scratch is not easier.
That got me thinking about the ways to implement modular widgets and I think it could be pretty nice to leverage some Widget registration system with each widget following some structure like
```
mywidget/main.js
mywidget/main.scss
mywidget/assets/***
```
With that structure all widgets could be put into ~/.config/ags/widgets, imported dynamically (let's say each widget exposes Register function). I think there are 2 benefits to that approach
1) readability - you know how widgets are supposed to be declared, where to look for them each time you see some widget used in configuration, but that's less important as I could implement that system for my own config, but readability arguably doesn't matter that much for dotfiles since I'll be the only one reading my trash code :P
2) sharing - that's what I think is the real benefit. If AGS encouraged people to declare widgets with certain structure it'd be possible to create easily sharable widgets. I think that without any effort it'd be very easy to use nix to import bunch of widgets from different places and just use them to declare your layout.

With system following that ideas I can imagine widget ecosystem created around AGS, which would lower barrier to entry and help avoid reinventing same things. With nix flakes and home manager module, there could be special attribute exposed like `extraWidgets` or `extraServices` (same idea as before really). That would split config into 2 parts - widget/service declaration part (imperative) and layout (more declarative).

**Comment by Aylur** [comment]
For this to work, we would need a stable version, which I am really close to releasing. This topic came up a few times on Discord now too, and I do like the idea and want a place to host user contributed widget implementations.
I have only though about a js and css file, which is really easy to use, its just an import and App.applyCss call, but an assets folder didn't cross my mind until now.
Once I'm happy with the state of ags I'll release a stable version, and then I'll get back to this.
